source/slang/slang-ir-inline.cpp

```diff
@@ -374,81 +381,138 @@ struct InliningPassBase
-                _cloneInstWithSourceLoc(callSite, env, builder, inst);
-                break;
-
-            case kIROp_Param:
-                // Parameters can be completely ignored in the single-block
-                // case, because they have all been replaced via `env`.
-                break;
-
-            case kIROp_ReturnVoid:
-                // A return with no operand can be ignored, since a return
-                // from the inlined call should just continue after the
-                // call site.
-                //
-                break;
-
-            case kIROp_ReturnVal:
-                // A return with a value is similar to `returnVoid` except
-                // that we need to note the (clone of the) value being
-                // returned, so that we can use it to replace the value
-                // of the original call.
-                //
-                returnedValue = findCloneForOperand(env, inst->getOperand(0));
-                break;
+                if (inst->getOp() == kIROp_Param)
+                {
+                    // Parameters in the first block can be completely ignored
+                    // because they have all been replaced via `env`.
+                    if (isFirstBlock)
+                    {
+                        continue;
+                    }
+                }
+
+                switch (inst->getOp())
+                {
+                default:
+                    // The default value is to clone the instruction using
+                    // the existing cloning infrastructure and the `env`
+                    // we have already set up.
+                    //
+                    // SourceLoc information is copied if there is appropriate data available.
+                    _cloneInstWithSourceLoc(callSite, env, builder, inst);
+                    break;
+
+                case kIROp_ReturnVoid:
+                    // A return with no operand is replaced with a branch into `afterBlock`
+                    // to return the control flow to the location after the original `call`.
+                    {
+                        auto returnBranch = builder->emitBranch(afterBlock);
+                        _setSourceLoc(returnBranch, inst, callSite);
+                    }
+                    break;
+
+                case kIROp_ReturnVal:
+                    // A return with a value is similar to `returnVoid` except
+                    // that we need to note the (clone of the) value being
+                    // returned, so that we can use it to replace the value
+                    // of the original call.
+                    //
+                    {
+                        auto returnBranch = builder->emitBranch(afterBlock);
+                        _setSourceLoc(returnBranch, inst, callSite);
+                        returnedValue = findCloneForOperand(env, inst->getOperand(0));
```

Comment:
I'm trying to think this through, and I'm not yet able to convince myself we can handle the more general class of functions being handled here.

Consider this case:

```
int f( int x )
{
    while( true )
    {
        while(true)
        {
            if( B() ) return C(); // single return site
            D();
        }
    }
}

int g(int y)
{
    while( E() )
    {
        F();
        y = f( y );
        if( G(y) ) break;
    }
    return y;
}
```

From what I can tell, this inlining pass is going to effectively replace the `return C();` statement with a branch to the code after the `y = f( y )` call site. This will *seem* to work if the code that is logically "after" the call site is restricted enough, but in a case like the above, we seem to end up with something like:

```
int g_optimized(int y)
{
    g_loop: while( E() )
    {
        F();
// begin inlined body of `f`
        outer_f_loop: while( true )
        {
            while(true)
            {
                if( B() ) {
                    y = C(); // assign directly to ressult... this part is easy...

                    // What do do here?!?!?!?!
                    //
                    // If we include the code after the call site directly,
                    // we need a multi-level `break` to implement the `break`
                    // from after the call site...
                    if( G(y) ) break g_loop;
                    //
                    // If instead we wanted to include the `if( G() )` code
                    // in its original location, we need to be able to do
                    break f_outer_loop;
                    //
                    // Either way we need the semantics of multi-level `break`!!! 
                }
                D();
            }
        }
// end inlined body of `f`
    }
    return y;
}
```

Am I missing something here?

Reply:
Right, we don't have a way to support multi-level break now.